analysis/dataset_definition_dm_reg.py

```diff
@@ -0,0 +1,86 @@
+from datetime import date, timedelta
+
+from databuilder.ehrql import Dataset, case, when
+from databuilder.tables.beta.tpp import (
+    clinical_events,
+    patients,
+    practice_registrations,
+)
+
+import codelists
+from variable_lib_helper import (
+    first_matching_event,
+    last_matching_event,
+    age_as_of,
+    died_as_of,
+    practice_registration_as_of,
+)
+
+# Payment Period End Date (PPED)
+index_date = date(2022, 3, 31)
+
+# Create dataset
+dataset = Dataset()
+
+# Define clinical events prior to index_date
+prior_events = clinical_events.take(clinical_events.date.is_on_or_before(index_date))
+
+# Define variable indicating whether a patient has died
+# Note: Although this is not in the business rules I'm including this as an
+# additional check
+has_not_died = ~died_as_of(index_date)
+
+
+# GMS reg status: Select patients who meet either of the criteria below:
+# Registered for GMS prior to or on the achievement date and did not
+# subsequently deregister from GMS (currently registered for GMS).
+# Registered for GMS prior to or on the achievement date and subsequently
+# deregistered from GMS after the achievement date (previously registered for
+# GMS). (i.e. patients who were registered for GMS on the achievement date).
+# TODO Note that this currently does not follow the qualifying criteria in the
+# business rules.
+practice_reg = practice_registration_as_of(index_date)
+gms_registration_status = practice_reg.exists_for_patient()
+
+# Define clinical data extraction criteria
+
+# Field number 4
+# PAT_AGE: The age of the patient in full years at the achievement date.
+dataset.pat_age = age_as_of(index_date)
+
+# Field number 6
+# DMLAT_DAT: Date of the most recent diabetes diagnosis up to and including the
+# achievement date.
+dataset.dm_lat_dat = last_matching_event(prior_events, codelists.dm_cod).date
```

Comment:
Just a quick drive-by comment.

I can see you're matching the column names to the source variable names, which is very sensible.

On the other hand, it'd make the code a lot more readable to use long, descriptive names, like `diabetes_most_recent_diagnosis` and `diabetes_most_recent_resolution`.

I feel like, on balance, the latter is more developer-friendly than the former, because we already reference the source variable in the comment.

What do you think?

Reply:
I agree that it could be helpful to use more meaningful variable names (`dm_lat_dat` vs `diabetes_most_recent_diagnosis_date`) provided we mention to the variable name from the source document.

however, I can see a couple problems that we need to consider and why I think we should stick to the variable names in the source document:
- it requires a little bit more time to write the initial code if we also have to come up with good variable names
- there are a lot of individual variables that we need to create for each indicator (66 for diabetes alone). when translating the QOF rules there are many references to these variables (e.g., date comparisons like `var1_dat` > `var2_dat`) and to me sticking with the variable names in the source document makes the translation part easier (because I dont have to remember how I renamed the variable)
- different people working on QOF indicators might end up with slightly different variable names for the same variable (some variables are shared across indicators, e.g., blood pressure). currently the QOF variables that are used across multiple  indicators have the same (not very descriptive) name. 
- if we have a reusable variable library someone new might not know what variable name we made up (although we could maintain a translation table somewhere) but it would always be easy to check the published QOF business rules
- currently the variable names have a naming protocol that is consistent and makes it easy to identify the codelist it is linked to (e.g., the variable `AMPL_DAT` refers to `AMPL_COD`)
- for the variables in the pull request it is _easy_ to come up with a short descriptive name, however most other variables describe a very specific thing, where it is harder to come up with a descriptive name, e.g.,
  - `AMPL_DAT`: Date of the most recent left foot amputation code up to and including the achievement date
  - `ACEDEC_COD`: The most recent date the patient chose not to receive ACE inhibitors up to and including the achievement date

@ccunningham101 and I went back and forth about this in our initial QOF studies and both believe following the docs has more pros than cons

Reply:
I understand and share @sebbacon's reservations about using the source variable names here, but agree with @milanwiedemann's rationale for doing so.